DR-2936 Enable copying files from  GCS to Azure

nmalfroy opened the pull request:
For more information, I did a write up of the approach of this PR here;
https://docs.google.com/document/d/1oIz3Czv89TAI8gBx37u7GJF4nqTORonTO4SI-B9yO-s

**Changes**
This PR introduces an exciting capability to TDR which is to support ingestion from GCS buckets into Azure datasets:

- In azure ingests, users can now specify, in addition to the existing Azure sources, `gs://` formatted URIs.
- The GCS URIs have been somewhat extended to support specifying a billing project with a query parameter named `userProject`.  This is needed for ingesting from a requests pays bucket.  So an ingest URI might look like: `gs://mybucket/path/to/y/blob?userProject=projectwiththemoney`
- Validation is performed on the source data for GCS data (no need for Azure since we work with signed URLs)
- On ingest, `gs://` paths are converted to 48 hour signed URLs (to give room for ingesting very large files) and that is fed into Azure's blob copy API that knows how to read that data...
- ...with the caveat that MD5s aren't copied and the signed URL is stored as a property on the blob (security risk) so, on completion of ingest, the MD5 is explicitly set (which unsets the source URL property)
- In order to simplify testing, this PR converts `BlobIOTestUtility` into an interface and has an Azure implementation (which already existed) and introduces a GCS version.  This is used to upload test data and clean up.

**Drive-by**
- Put in a fix that should make the storage accounts get cleaned up more aggressively in our tests.  That's handled in the integration test commit at the end
- Not longer need to specify the billing account to ingest records! If it isn't specified, we get it from the dataset (this was driving me insane)

**Testing**
- This is a big enough change that there is all manner of testing: unit, connected and integration
- A lot of manual testing was also done to test scale.  I tested copying files up to 40GB (26 minutes, for the record) though the biggest file we expect will be 1.3 TB

nmalfroy committed:
Add a utility method to work with URIs

nmalfroy committed:
Drive-by fix to fix java doc

nmalfroy committed:
GcsPdao changes

- Add ability to opt out of adding the user's pet service account to the dataset's
  project when verifying access.
- Add a method to extract the billing project ID from a GCS URI
- Add ability for getBlobFromGsPath to not require a billing project
- Create a non-static getBlobFromGsPath (getBlobFromGsPathNs) to support easier testing

nmalfroy committed:
Multicloudify test utility to create files on Azure and GCS

nmalfroy committed:
Add support for the copier utility method to support GCS sourced files

nmalfroy committed:
Make sure that MD5s are recorded after copy

nmalfroy committed:
Drive by: use the default billing id of the dataset when ingesting metadata

...when no billing id is specified

nmalfroy committed:
Add GCS source data to AzureIntegrationTest

Note: also added better handling for making sure that we properly delete storage account when we're done

pshapiro4broad commented on src/main/java/bio/terra/common/UriUtils.java:
```diff
@@ -0,0 +1,70 @@
+   * @return a new uri string any reference to the query parameter removed
+   */
+  public static String omitQueryParameter(String uri, String parameter) {
+    URI initUri = toUri(uri);
+    List<NameValuePair> newQueryParams =
+        URLEncodedUtils.parse(toUri(uri), Charsets.UTF_8).stream()
+            .filter(p -> !Objects.equals(p.getName(), parameter))
+            .toList();
+    return new URIBuilder(initUri).setParameters(newQueryParams).toString();
```
Have you considered using spring `UriComponents` for this kind of operation? for omitting a query parameter, something like this should work:
```
    return UriComponentsBuilder.fromUriString(uri).replaceQueryParam(parameter).build().toString();
```

To get a query param, you'd write something like:
```
  List<String> value = UriComponentsBuilder.fromUriString(uri).build().getQueryParams().get(parameter);
```

nmalfroy replied:
oooh!  today I learned, thanks!  I'll change to this.  thanks!

nmalfroy replied:
ah sadness, I reverted it because it breaks URIs in certain cases (and I don't really care that much about drilling in...I don't really trust this method anymore)

nmalfroy replied:
actually, nevermind...might be something I did :(

pshapiro4broad commented on src/main/java/bio/terra/service/common/gcs/GcsUriUtils.java:
```diff
@@ -13,7 +13,7 @@ public final class GcsUriUtils {
    * Parse a Google Cloud Storage URI into its component pieces
    *
-   * @param uri of type gs://<bucket_name>/<file_path_inside_bucket>
+   * @param uri of type gs://[bucket_name]/[file_path_inside_bucket]
```
If you wanted to keep the `<` `>` you could instead wrap in `@code`
```suggestion
   * @param uri of type {@code gs://<bucket_name>/<file_path_inside_bucket>}
```

nmalfroy replied:
Eh, yeah, I started with that and I honestly don't care about having brackets

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/azure/blobstore/AzureBlobStorePdao.java:
```diff
@@ -209,10 +273,23 @@ public Stream<String> getBlobsLinesStream(
+            .map(p -> UriUtils.getValueFromQueryParameter(p, USER_PROJECT_QUERY_PARAM))
+            .filter(Objects::nonNull)
+            .distinct()
+            .toList();
```
~You could use `count()` here instead of `toList()`~ nevermind I see you need the list contents below

nmalfroy replied:
eh, I need the list below

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/azure/util/BlobContainerCopier.java:
```diff
@@ -22,6 +38,9 @@ public class BlobContainerCopier {
   private final BlobContainerClientFactory destinationClientFactory;
+  private static final String USER_PROJECT_QUERY_PARAM = "userProject";
+  // Signing URL for 48 hours to handle large file transfers
+  private static final int SIGNED_URL_DURATION_MINUTES = 60 * 48;
```
You can use `TimeUnit` to handle `* 60` for you
```suggestion
  private static final long SIGNED_URL_DURATION_MINUTES = TimeUnit.HOURS.toMinutes(48);
```
Or use 48 here and pass in `TimeUnit.HOURS` below.

nmalfroy replied:
good call.

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/azure/util/BlobContainerCopier.java:
```diff
@@ -118,17 +137,22 @@ public List<BlobCopySourceDestinationPair> getSourceDestinationPairs() {
-    BlobUrlParts blobUrlParts = BlobUrlParts.parse(sourceBlobUrl);
+  private BlobContainerCopySyncPoller beginCopyOperationUsingUrl() {
+    String sourceBlobName;
+    if (sourceBlobUrl.startsWith("gs://")) {
```
We really have a lot of `.startsWith("gs://")` calls now. Maybe it should be a function in `GcsUriUtils`, like `GcsUriUtils.isGsPath()`?

nmalfroy replied:
Hmm, yeah, good point

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/azure/blobstore/AzureBlobStorePdao.java:
```diff
@@ -209,10 +273,23 @@ public Stream<String> getBlobsLinesStream(
+    }
+
+    gcsPdao.validateUserCanRead(
+        gsPaths, cloudEncapsulationIds.stream().findFirst().orElse(null), user, false);
```
I think you want `orElseThrow()` here since `cloudEncapsulationId` can't be null (can it?).
```suggestion
        gsPaths, cloudEncapsulationIds.stream().findFirst().orElseThrow(), user, false);
```

Also maybe change `size() > ` to `size() == 1` to preflight this check. Then you can replace the `stream` with `cloudEncapsulationIds.get(0)`.

nmalfroy replied:
It can actually be null.  That means no project to bill to (which for loading into Azure is good)

pshapiro4broad replied:
OK. It will result in a key entry that contains `null` since it's used in `validateUserCanRead()` like this:
```
String.format("%s%s%s", user.getToken(), PSA_SEPARATOR, cloudEncapsulationId)
```

If `null` is expected, maybe the `format()` should be changed to check for that case?

nmalfroy replied:
Ah yeah, great point.  I suppose you can't name a project null so it's moot but I agree, that's gross.  I'll fix it

nmalfroy committed:
Add isGcsUri helper method

okotsopoulos commented on src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java:
```diff
@@ -303,6 +304,29 @@ public Blob createGcsFile(BlobId blobId, String projectId) {
+      List<String> sourcePaths,
+      String cloudEncapsulationId,
+      AuthenticatedUserRequest user,
+      boolean addPSAToDatasetProject) {
```
As a heads up, I'm also modifying this method as part of skipping permission validation for dedicated SA datasets!

I bring it up because I ended up passing the whole dataset to `validateUserCanRead`, and adding a helper method `Dataset.shouldValidateIngesterFileAccess()`.  That way the logic stays close to the dataset itself.  I can throw up a draft PR if it helps.

What do you think about this approach instead, especially as we'll be supporting multiple conditional behavior mods?

nmalfroy replied:
Ooh, I like that.  If your goes in first, I'm happy to convert to use your version

okotsopoulos replied:
Awesome! Lights a fire under me to get this out, ha! (I don't mind rebasing off of yours though.)

nmalfroy replied:
haha, no fire needed

okotsopoulos replied:
As discussed in stand-up, this PR is likely to go in before [mine](https://github.com/DataBiosphere/jade-data-repo/pull/1453). I will rebase and pull the dataset through :)

okotsopoulos commented on src/main/java/bio/terra/common/UriUtils.java:
```diff
@@ -0,0 +1,70 @@
+/** Handy methods for working with URIs */
+public class UriUtils {
+
+  private UriUtils() {}
```
Ah, I figured out how to avoid needing a no-op constructor (if you'd like): annotating the class as a `@Component` allows you to skip it.

pshapiro4broad replied:
I would recommend against using `@Component` as a way to express a utils class, since making it a component means that it will be instantiated as part of the application context (when what is really wanted is to never instantiate it). It's also not what spring `component` is supposed to represent.

If you want to avoid adding a no-op constructor (and declaring the class final, etc) I'd recommend using the enum pattern instead. In that case you define an enum with no members:
```
public enum UriUtils {
    ;

   [...]
```

And no other code or annotations are needed. This also works if you want to represent a singleton object; in that case you can declare a single enum member which will be the singleton instance.

nmalfroy replied:
I'm gonna leave as a class for consistency

okotsopoulos commented on src/main/java/bio/terra/service/common/gcs/GcsUriUtils.java:
```diff
@@ -118,4 +118,15 @@ public static String makeHttpsFromGs(String gspath, String userProject) {
+
+  /**
+   * Performs rudimentary test on a potential gcs uri to see if it might be valid (note: does not
+   * confirm the validity of the gcs path
```
Nit: missing a trailing parens.

okotsopoulos commented on src/main/java/bio/terra/service/filedata/azure/util/BlobContainerCopier.java:
```diff
@@ -22,6 +38,9 @@ public class BlobContainerCopier {
   private final BlobContainerClientFactory destinationClientFactory;
+  private static final String USER_PROJECT_QUERY_PARAM = "userProject";
+  // Signing URL for 48 hours to handle large file transfers
+  private static final long SIGNED_URL_DURATION_MINUTES = TimeUnit.HOURS.toMinutes(48);
```
Would this be something that we'd prefer to have config control over?

nmalfroy replied:
Yeah, that's a great call

nmalfroy replied:
Hmm...I'm actually going to leave it for now. The way the copier code is set up makes it really gross to add and we've already got the pattern of hard coding these values for this class

okotsopoulos commented on src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java:
```diff
@@ -352,11 +380,22 @@ public void validateUserCanRead(
+        // This is a potential failure mode for permissions checking: not being able to make the
+        // permissions check call at all
+        if (e.getCode() == HttpStatus.SC_FORBIDDEN) {
+          permissions = null;
```
Is this explicit null set needed?  Any reason that you didn't write this block as:

```
if (e.getCode() != HttpStatus.SC_FORBIDDEN) {
  throw e;
}
```

pshapiro4broad replied:
If you don't set `null` here, you get an error since `permissions` is no longer initialized before use below. That could be avoided by initializing `permissions` when it's declared, but then you're writing to it twice.

In any case I would recommend using `List.of()` instead of `null` to avoid the null check below.
```suggestion
          permissions = List.of();
```

Another approach is to do the check of `permissions.equals(List.of(true))` after line 387 above and return early if true. Then you no longer need `permissions` outside the scope of the `try` and the `if` can be removed.

nmalfroy replied:
Ah derp, that's a great point

okotsopoulos commented on src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java:
```diff
@@ -674,19 +713,41 @@ public void blobAclUpdates(
   }
 
+  /**
+   * Non-static version of {@link GcsPdao#getBlobFromGsPath(Storage, String, String)} to enable
```
Ah dang, nice to be able to mock this but this smells funny.  Do we need to maintain the existing static version?

pshapiro4broad replied:
Seems like mocking `Storage` would be better if it's not too hard to do.

nmalfroy replied:
Hmm...lemme see if I can do that.  I agree it's less gross if we can avoid the mixed static / non-static thing

nmalfroy replied:
k, mocked out storage. I like it way better.  thanks for the suggestion!

okotsopoulos commented on src/test/java/bio/terra/integration/AzureIntegrationTest.java:
```diff
@@ -1046,14 +1065,14 @@ sourceFile, getSourceStorageAccountPrimarySharedKey()))
 
     // Make sure that any failure in tearing down is presented as a test failure
-    blobIOTestUtility.deleteContainers();
+    azureBlobIOTestUtility.teardown();
```
I don't think this is necessary as we call this teardown method as part of test teardown.  (Ditto throughout this test suite.)

nmalfroy replied:
I think that was there to get the error to show up as a test failure.  I look into handling this in a followup if that's ok

okotsopoulos commented on src/test/java/bio/terra/service/filedata/azure/blobstore/AzureBlobStorePdaoTest.java:
```diff
@@ -175,6 +189,46 @@ public void testCopyFileWithSas() {
+            AZURE_STORAGE_ACCOUNT_RESOURCE,
+            TEST_USER);
+    assertThat("output is expected", fsFileInfo, samePropertyValuesAs(expectedFileInfo));
+    verify(gcsPdao, times(1)).getBlobFromGsPathNs(any(), eq(SOURCE_GCS_PATH), eq(null));
```
If you wish, you can omit `times(1)` since that's the default verification.

nmalfroy replied:
I ended up just removing the verify since just having a successful call is validation enough

okotsopoulos reviewed (approved):
Looks good!  I left a few comments.  (Aside: whew, 48h seems to be quite a long-lived URL!)

samanehsan commented on src/test/java/bio/terra/common/UriUtilsTest.java:
```diff
@@ -0,0 +1,74 @@
+        "Invalid URI",
+        () -> UriUtils.getValueFromQueryParameter(badUri, "q"));
+
+    TestUtils.assertError(
```
I think this is a duplicate of the assertion right above it

nmalfroy replied:
doh!  It sure is

samanehsan reviewed (approved):
This looks good to me! Thanks for including a write-up on your approach here!

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java:
```diff
@@ -674,19 +713,41 @@ public void blobAclUpdates(
+   * @return the value of the userProject query parameter or null
+   */
+  public static String getProjectIdFromGsPath(String gspath) {
+    return UriUtils.getValueFromQueryParameter(gspath, "userProject");
```
Could this use a named constant?
```suggestion
    return UriUtils.getValueFromQueryParameter(gspath, GcsConstants.USER_PROJECT_QUERY_PARAM);
```

nmalfroy replied:
Gah, great catch.  I thought I'd caught all of them

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/google/gcs/GcsConstants.java:
```diff
@@ -0,0 +1,13 @@
+   * accessing the blob. Note: this is not a Google-recognized query parameter so it must be removed
+   * when accessing the file using Google clients
+   */
+  public static final String USER_PROJECT_QUERY_PARAM = "userProject";
```
This already exists in two other places. If it's the same concept, could they all use the same constant? I see `DrsService.USER_PROJECT_QUERY_PARAM` and `BlobContainerCopier.USER_PROJECT_QUERY_PARAM`

nmalfroy replied:
They're subtly different.  They happen to be the same values but one is the one formally recognized by Google and used in their APIs, the other is ours so I wanted to keep them separate

pshapiro4broad commented on src/main/java/bio/terra/common/UriUtils.java:
```diff
@@ -0,0 +1,70 @@
+   * @return the value of the query parameter or null. If there are multiple instances of the query
+   *     parameter, return the value as a comma-separated list
+   */
+  public static String getValueFromQueryParameter(String uri, String parameter) {
```
In the sprit of YAGNI I'd like to point out that both `getValueFromQueryParameter()` and `omitQueryParameter()` are only invoked with `GcsConstants.USER_PROJECT_QUERY_PARAM` as a value for `parameter`. Is a general method needed here, or would one that is specific to the current use case be sufficient?

nmalfroy replied:
I wanted a general method since this is something I can see us needing and I wanted to just have it out there for the next person

nmalfroy committed:
Followons

- Remove non-static version of GcsPday.getBlobFromGsPath
- Consolidate userProject parameter constants
- Test cleanup
